app/grandchallenge/components/forms.py

```diff
@@ -151,8 +164,63 @@ def _get_image_field(self, *, interface, current_value):
         )
 
     def _get_file_field(self, *, interface, values, current_value):
-        return self._get_default_field(
-            interface=interface, current_value=current_value
+        if current_value:
+            if isinstance(current_value, ComponentInterfaceValue):
+                # this happens on initial form load when there already is a CIV for
+                # the interface
+                return self._get_select_upload_widget_field(
+                    interface=interface,
+                    values=values,
+                    current_value=current_value,
+                )
+            else:
+                # on form submit, current_value either is the pk of an existing CIV
+                # or the UUID of a new UserUpload object
+                type, value = current_value.split("_")
+                if type == "uuid":
+                    return self._get_default_field(
+                        interface=interface, current_value=value
+                    )
+                elif type == "civ":
+                    # value can be None (when user selects '---' in the select widget)
+                    if (value and int(value) in values) or not value:
+                        return self._get_select_upload_widget_field(
+                            interface=interface,
+                            values=values,
+                            current_value=ComponentInterfaceValue.objects.get(
+                                pk=value
+                            )
+                            if value
+                            else None,
+                        )
```

Comment:
This is hard to understand now as there are lots of packed clauses, can we try to make it easier on our future selves? Maybe split it out a bit? This is what I think is going on, but I am unsure:


```python
try:
    if int(value) in values:
        current_value = ComponentInterfaceValue.objects.get(
              pk=value
        )
    else:
        # The selected CIV does not belong to the set of values
        raise PermissionDenied
except TypeError:
    current_value = None

return self._get_select_upload_widget_field(
                             interface=interface,
                             values=values,
                             current_value=current_value,
                         )
```

Reply:
That is exactly what is going on, except that `int(None)` returns a `ValueError` not a TypeError. Thanks!

Reply:
Are you sure?

```
>>> int(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

Reply:
Also made a change that puts the exception handling next to the line we expect to find it so that we don't have huge try blocks.

Reply:
Ah, sorry for the confusion, since we're parsing `value` from a string with `current_value.split("_")`, `value` won't be `None` but `''` instead, and  `int('')` raises a ValueError.

Reply:
Ah okay, then update the comment about it as it's misleading.